source/d3d9/runtime_d3d9.cpp

```diff
@@ -378,19 +484,149 @@ namespace reshade::d3d9
 
 		if (depthstencil != nullptr)
 		{
-			if (depthstencil == _depthstencil_replacement)
+			if (depthstencil == get_depthstencil_replacement())
 			{
+				// for the next tables, we need to get back to the original depthstencil ref
 				depthstencil = _depthstencil;
 			}
 
-			const auto it = _depth_source_table.find(depthstencil.get());
+			if (!_preserve_depth_buffer)
+			{
+				// original depth source table update
+				const auto it = _depth_source_table.find(depthstencil.get());
+
+				if (it != _depth_source_table.end())
+				{
+					it->second.drawcall_count = _drawcalls;
+					it->second.vertices_count += vertices;
+				}
+			}
+			else
+			{
+				_preserve_drawcalls++;
+				_preserve_vertices += vertices;
+
+				depthstencil = get_depthstencil_replacement();
+			}
+		}
+	}
+	void runtime_d3d9::before_clear(com_ptr<IDirect3DSurface9> depthstencil)
+	{
+		// early rejection
+		if (depthstencil == nullptr || !_preserve_depth_buffer || depthstencil != get_depthstencil_replacement())
+			return;
+
+		const auto it = _depth_buffer_table.find(_clear_buffer_idx);
+
+		if (it != _depth_buffer_table.end())
+		{
+			it->second.drawcall_count = _preserve_drawcalls;
+			it->second.vertices_count = _preserve_vertices;
+		}
+
+		const auto it2 = _depth_clearing_table.find(_clear_idx);
+
+		if (it2 != _depth_clearing_table.end())
+		{
+			it2->second.drawcall_count = _preserve_drawcalls;
+			it2->second.vertices_count = _preserve_vertices;
+		}
+
+		_preserve_drawcalls = 0;
+		_preserve_vertices = 0;
+
+		if (_auto_preserve && _multi_depthstencil)
+		{
+			// when multiple depth buffer textures are set (one per depth buffer clearance), redefine the best depth buffer clearing instance
+			_preserve_starting_index = get_best_preserve_starting_index(true);
+		}
+
+		if (is_preserved())
+		{
+			// if the current depth buffer replacement texture has to be preserved, replace the depthStencilSurface ref with a dummy one, so the depth buffer replacement texture will not be cleared
+			_device->SetDepthStencilSurface(_depthstencil.get());
+		}
+	}
+	void runtime_d3d9::after_clear(com_ptr<IDirect3DSurface9> depthstencil)
```

Comment:
I don't think this has to be this complicated. Instead of splitting this code up into two functions `before_clear` and `after_clear`, how about using a single one that returns a bool which specifies whether to execute the clear operation or not. If it is false, return from `Direct3DDevice9::Clear` without calling the original device `Clear`. This way you are saving a lot of duplicated checks and there is no need to replace the currently set depth stencil surface with another one before clear.